### PR TITLE
fix: return itself when getOutboundNodes from memoization Hit steps/DAG. Fixes: #7873

### LINKED ISSUE #7873: A memoized steps hit, the following nodes not displayed in the UI
## Summary

When a memoized step with steps template is hit by a ConfigMap record, the nodes following this step are not displayed in the UI (despite they run expectedly). The behavior might result from that the memoized step has no children.

## Diagnostics

The smallest workflow that reproduces the bug looks like

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: memoized-bug-
spec:
  entrypoint: main
  templates:
  - name: main
    steps:
    - - name: hello-steps
        template: memoized
    - - name: whatever
        template: hello

  - name: memoized
    outputs:
      parameters:
      - name: msg
        valueFrom:
          parameter: "{{steps.hello-step.outputs.result}}"
    steps:
    - - name: hello-step
        template: hello
    memoize:
      key: "memoized-bug-steps-0"
      cache:
        configMap:
          name: my-config

  - name: hello
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo Hello"]
```

When the workflow is submitted for the first time, the step `hello-steps` is stored in the ConfigMap. And in the UI, the workflow looks like
workflow-1
Everything is OK. But when the workflow is submitted for the second time, the step `hello-steps` is loaded from the ConfigMap, the following step `whatever` is not displayed in the UI
workflow-2

In the detail of the second workflow
```bash
$ argo get memoized-bug-nmvgq -n argo -o yaml
...
    memoized-bug-nmvgq-667942788:
      boundaryID: memoized-bug-nmvgq
      displayName: hello-steps
      finishedAt: "2022-02-13T05:32:52Z"
      id: memoized-bug-nmvgq-667942788
      memoizationStatus:
        cacheName: my-config
        hit: true
        key: memoized-bug-steps-0
      name: memoized-bug-nmvgq[0].hello-steps
      outputs:
        parameters:
        - name: msg
          value: Hello
      phase: Succeeded
      startedAt: "2022-02-13T05:32:52Z"
      templateName: memoized
      templateScope: local/memoized-bug-nmvgq
      type: Steps
...
    memoized-bug-nmvgq-2677016171:
      boundaryID: memoized-bug-nmvgq
      displayName: whatever
      finishedAt: "2022-02-13T05:33:05Z"
      hostNodeName: minikube
      id: memoized-bug-nmvgq-2677016171
      name: memoized-bug-nmvgq[1].whatever
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: memoized-bug-nmvgq/memoized-bug-nmvgq-2677016171/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 4
        memory: 2
      startedAt: "2022-02-13T05:32:52Z"
      templateName: hello
      templateScope: local/memoized-bug-nmvgq
      type: Pod
```
The step `hello-steps` does not have children which may cause the step `whatever` not displayed in the UI, despite the step `whatever` succeeded.

---
Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by shuangkun**
Fixes #7873

### Motivation

The first step "hello-steps" should have one child (The SgNode of "whatever")，but when “hello-steps” hit memoize，it has no child because of getOutboundNodes from the node return nil. Prevents the next sgNode from becoming the child of the previous sgNode’s outboundNodes.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: memoized-bug-
spec:
  entrypoint: main
  templates:
  - name: main
    steps:
    - - name: hello-steps
        template: memoized
    - - name: whatever
        template: hello

  - name: memoized
    outputs:
      parameters:
      - name: msg
        valueFrom:
          parameter: "{{steps.hello-step.outputs.result}}"
    steps:
    - - name: hello-step
        template: hello
    memoize:
      key: "memoized-bug-steps-0"
      cache:
        configMap:
          name: my-config

  - name: hello
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo Hello"]
```

### Modifications

return itself when getOutboundNodes from memoization Hit steps/DAG.

### Verification

Local test and unit tests.

### CODE REVIEW

**Comment by shuangkun** [comment]
Hi,Joibel @Joibel  I know you are very familiar with memoize, can you help review this, when a node hits the cache, its node id can be returned when getOutboundNodes.  Thank you so much！

**Comment by Joibel** [comment]
Happy to look at it. I am at Kubecon/Argocon this week, so not sure whether I will get a chance until next week.

Can you get the test suite to ✅?

**Comment by shuangkun** [comment]
> Happy to look at it. I am at Kubecon/Argocon this week, so not sure whether I will get a chance until next week.
> 
> Can you get the test suite to ✅?

Thanks, enjoy your conference!  My question is not urgent, you can take a look at my question later when you have time.  The test suite is OK.

File: workflow/controller/operator_test.go [hunk 1]
```diff
@@ -5383,6 +5383,171 @@ func TestConfigMapCacheLoadOperateNoOutputs(t *testing.T) {
...
+
+	woc := newWorkflowOperationCtx(wf, controller)
+	woc.operate(ctx)
+	makePodsPhase(ctx, woc, apiv1.PodSucceeded)
+	woc.operate(ctx)
+
+	assert.Equal(t, wfv1.WorkflowSucceeded, woc.wf.Status.Phase)
+
+	for _, node := range woc.wf.Status.Nodes {
+		if node.DisplayName == "hello-steps" {
```
[the remark this answers is not in the document]
Comment (shuangkun) [reply workflow/controller/operator_test.go:5500]: Thanks! I add it.

**Review by Joibel — APPROVED** [review]
LGTM. @agilgur5 @isubasinghe @terrytangyuan could you do the magic please

**Review by isubasinghe — APPROVED** [review]

**Comment by agilgur5** [comment]
Backported cleanly to `release-3.5` as 4e7d471c0

**Pull request merged**